runtime/src/accounts_background_service.rs

```diff
@@ -0,0 +1,160 @@
+
+                // Check to see if there were any requests for snapshotting banks
+                // < the current root bank `bank` above.
+
+                // Claim: Any snapshot request for slot `N` found here means that cleanup has
+                // been called at most on slot `N`
+                //
+                // Proof: Assume for contradiction that we find a snapshot request for slot `N` here,
+                // but cleanup has already happened on some slot `M > N`.
+                //
+                // Then that means on some previous iteration of this loop, we found `bank_forks.root_bank() == M`,
+                // but did not see the snapshot for `N`. However, this is impossible because BankForks.set_root()
+                // will always flush the snapshot request for `N` to the snapshot request channel
+                // before setting a root `M > N`.
+                let snapshotted_slot = snapshot_items.as_ref().and_then(|snapshot_items| {
+                    let SnapshotItems {
+                        snapshot_config,
+                        snapshot_request_receiver,
+                        accounts_package_sender,
+                    } = snapshot_items;
+                    Self::process_snapshot_requests(
+                        snapshot_config,
+                        snapshot_request_receiver,
+                        accounts_package_sender,
+                    )
+                });
+
+                if let Some(snapshotted_slot) = snapshotted_slot {
+                    // Safe, see proof above
+                    assert!(last_cleaned_slot <= snapshotted_slot);
+                    last_cleaned_slot = snapshotted_slot;
+                } else {
+                    consumed_budget = bank.process_stale_slot_with_budget(
+                        consumed_budget,
+                        SHRUNKEN_ACCOUNT_PER_INTERVAL,
+                    );
+                }
+
+                sleep(Duration::from_millis(INTERVAL_MS));
+            })
+            .unwrap();
+        Self { t_background }
+    }
+
+    pub fn join(self) -> thread::Result<()> {
+        self.t_background.join()
+    }
+
+    // Returns the latest rquested snapshot slot, if one exists
+    pub fn process_snapshot_requests(
+        snapshot_config: &SnapshotConfig,
+        snapshot_request_receiver: &SnapshotRequestReceiver,
+        accounts_package_sender: &AccountsPackageSender,
+    ) -> Option<Slot> {
+        snapshot_request_receiver
+            .try_iter()
+            .last()
+            .map(|snapshot_request| {
+                let SnapshotRequest {
+                    snapshot_root_bank,
+                    status_cache_slot_deltas,
+                } = snapshot_request;
+
+                snapshot_root_bank.process_stale_slot_with_budget(0, SHRUNKEN_ACCOUNT_PER_INTERVAL);
+                // Don't clean the slot we're snapshotting becaue it may have zero-lamport
```

Comment:
typo: `because`

Reply:
Also, how about adding test for this bad hash mismatch?

Reply:
(the test will justify the existence of the new `max_root` thing, which starts to be passed here and there.)

Reply:
Test added here: https://github.com/solana-labs/solana/pull/12194/commits/e7c1ce44aac87b64b7c10e56e00c9ebf58853d86